File: x86_trace_generator/pin/bindings/glue/pinglueserver/tests/Scripts/GetSourceLocation.py
#
# Copyright (C) 2024-2024 Intel Corporation.
# SPDX-License-Identifier: MIT
#

import pin

FILE_TO_FIND = "hello"
info = open("script_GetSourceLocation.out", "w")

def image(img):
    global FILE_TO_FIND
    if not pin.IMG_IsMainExecutable(img):
        return
    info.write(f"Image name = {pin.IMG_Name(img)}\n")

    sec = pin.IMG_SecHead(img)
    while(pin.SEC_Valid(sec)):
        rtn = pin.SEC_RtnHead(sec)
        
        while(pin.RTN_Valid(rtn)):
            pin.RTN_Open(rtn)
            
            ins = pin.RTN_InsHead(rtn)
            while(pin.INS_Valid(ins)):
                res = pin.PIN_GetSourceLocation(pin.INS_Address(ins))
                if FILE_TO_FIND in res[2]:
                    info.write(f"column: {res[0]}, line: {res[1]}, srcfile: {res[2]}\n")         
            
                ins = pin.INS_Next(ins)
            
            pin.RTN_Close(rtn)
            rtn = pin.RTN_Next(rtn)
        sec = pin.SEC_Next(sec)    

pin.IMG_AddInstrumentFunction(image)

File: x86_trace_generator/pin/bindings/glue/pinglueserver/tests/NativeTools/GetSourceLocation.cpp
/*
 * Copyright (C) 2024-2024 Intel Corporation.
 * SPDX-License-Identifier: MIT
 */

#include <stdlib.h>
#include <stdint.h>
#include <stdio.h>
#include <iostream>
#include <fstream>
#include <map>
#include <vector>
#include <string>
#include "pin.H"
using std::cerr;
using std::cout;
using std::endl;
using std::ios;
using std::string;

string FILE_TO_FIND = "hello";

std::ofstream out;

void InstImage(IMG img, void* v)
{
    if (!IMG_IsMainExecutable(img)) return;

    out << "Image name = " << IMG_Name(img) << endl;

    for (SEC sec = IMG_SecHead(img); SEC_Valid(sec); sec = SEC_Next(sec))
    {
        for (RTN rtn = SEC_RtnHead(sec); RTN_Valid(rtn); rtn = RTN_Next(rtn))
        {
            RTN_Open(rtn);

            for (INS ins = RTN_InsHead(rtn); INS_Valid(ins); ins = INS_Next(ins))
            {
                int column, lineno;
                string srcfile = "";
                PIN_GetSourceLocation(INS_Address(ins), &column, &lineno, &srcfile);
                if (srcfile.find(FILE_TO_FIND) != srcfile.npos)
                {
                    out << "column: " << column << ", line: " << lineno << ", srcfile: " << srcfile << endl;
                }
            }
            RTN_Close(rtn);
        }
    }
}

int main(int argc, char** argv)
{
    PIN_InitSymbolsAlt(SYMBOL_INFO_MODE(UINT32(IFUNC_SYMBOLS) | UINT32(DEBUG_OR_EXPORT_SYMBOLS)));

    if (PIN_Init(argc, argv))
    {
        cerr << "usage..." << endl;
        return EXIT_FAILURE;
    }

    out.open("GetSourceLocation.out");
    out.setf(ios::showbase);
    IMG_AddInstrumentFunction(InstImage, 0);

    PIN_StartProgram();

    return 0;
}
